unsloth/models/llama.py

```diff
@@ -113,6 +113,46 @@ def original_apply_o(self, X):
 # SDPA has GQA internally
 SDPA_HAS_GQA = "enable_gqa" in scaled_dot_product_attention.__doc__
 
+
+def _prepare_model_for_qat(model: torch.nn.Module, qat_scheme: str) -> torch.nn.Module:
+    """
+    Apply QAT + LoRA during fine-tuning.
+
+    On a high level, this means fake quantizing the base (frozen) model during LoRA training.
+    Fake quantization refers to simulating quantization numerics in high precision (e.g. bf16).
+    This helps mitigate quantization degradations when the model is quantized after training.
+
+    For more details: https://dev-discuss.pytorch.org/t/speeding-up-qat-by-1-89x-with-lora/2700
+    """
+    try:
+        from torchao.quantization import (
+            Float8DynamicActivationFloat8WeightConfig,
+            Float8DynamicActivationInt4WeightConfig,
+            PerRow,
+            quantize_,
+        )
+        from torchao.quantization.qat import QATConfig
+    except ImportError as e:
+        print(
+            "Please install torchao nightly for the latest QAT features:\n"
+            "  pip install --pre torchao --index-url https://download.pytorch.org/whl/nightly/cu126"
+        )
+        raise e
+    pass
+    filter_fn = None
+    if qat_scheme == "fp8-int4":
+        group_size = 128
+        base_config = Float8DynamicActivationInt4WeightConfig(group_size=group_size)
+        filter_fn = lambda m, _: isinstance(m, torch.nn.Linear) and m.in_features >= group_size
```

Comment:
we could skip this in the config handler itself

Reply:
we also plan to remove the group_size arg for now since there is no benefit of making it larger according to Josh